### DISCUSSION TITLE
IState<T> update property inside object doesn`t work

### DISCUSSION THREAD

**Discussion by limeniye** [unoplatform/uno.extensions#1163] [Q&A]
I have AuthorizationCredentials type:
```C#
public enum AutorizationMode
{
    SignIn,
    SignUp
}

public record AuthorizationCredentials
{
    public AutorizationMode AutorizationMode { get; init; }

    public string? UserName { get; init; }

    public string? Password { get; init; }

    public string? RepeatedPassword { get; init; }
}
```

I`m try to notify my view that AutorizationMode property was changed.
```C#
public IState<AuthorizationCredentials> Credentials => State<AuthorizationCredentials>.Empty(this);

public async ValueTask SomeTestMethod(CancellationToken ct)
{
    await Credentials.UpdateValue(opt => opt.Map(credintials => credintials with { AutorizationMode = AutorizationMode.SignUp}), ct);
}
```

```XML
<PasswordBox x:Name="repeatePassowrdTextBox">
    <i:Interaction.Behaviors>
        <core:DataTriggerBehavior Binding="{Binding Credentials.AutorizationMode, Mode=OneWay, Converter={StaticResource EnumToIntConverter}}" Value="0">
            <core:ChangePropertyAction TargetObject="{Binding ElementName=repeatePassowrdTextBox}" PropertyName="Visibility" Value="Collapsed"/>
        </core:DataTriggerBehavior>
    </i:Interaction.Behaviors>
</PasswordBox>
```

But the breakpoint in mine EnumToIntConverter doesn't work after UpdateValue.
Am I doing something wrong?

**Comment by francoistanguay** [comment]
Are your bindings working (besides the changes not working?) ?
That state object must be contained in a class (typically a Model), for which the Reactive Source Generator is generating a "Bindable" version of it and this is what should be set in the DataContext of your View.

Are you able to share a repro? A bit more details about the surrounding code? Whether or not it's using Uno Navigation?

Thanks!

**Reply by limeniye** [reply]
@francoistanguay, yes, I'm using Uno navigation. 

Passowrd and Login binding works well because I don't use some converters and don't change values from ViewMode code. 

- [App.xaml.hosts](https://github.com/ligric/ligricui/blob/features/nhibernate-signup/src/Ligric.UI.Shared/App.xaml.host.cs#L70)
- [AuthorizationViewModel](https://github.com/ligric/ligricui/blob/features/nhibernate-signup/src/Ligric.UI.ViewModels/Presentation/AuthorizationViewModel.cs#L17)
- [AuthorizationPage](https://github.com/ligric/ligricui/blob/features/nhibernate-signup/src/Ligric.UI.Shared/Views/AuthorizationPage.xaml#L53-L62)
- [EnumToIntConverter](https://github.com/ligric/ligricui/blob/features/nhibernate-signup/src/Ligric.UI.Shared/Converters/EnumToIntConverter.cs)

**Reply by dr1rrb** [reply]
Hey @limeniye,

I think your issue is because you have created an **Empty** state : [`State<AuthorizationCredentials>.Empty(this);`](https://github.com/ligric/ligricui/blob/features/nhibernate-signup/src/Ligric.UI.ViewModels/Presentation/AuthorizationViewModel.cs#L17). `Empty` means that you are creating a state which **does not have any value**, i.e. `Option.None<AuthorizationCredentials>()` (if you used a `FeedView` to render it, you have been in the empty visual state).

Then when you try to update it, you are using [`Option.Map()`](https://github.com/unoplatform/uno.extensions/blob/main/src/Uno.Extensions.Core/OptionExtensions.cs#L36) which is a helper that will apply the _projection_ only if the given option do have a value (i.e. `Option.Some`), and will keep the previous option's type otherwise. So in your case you are staying in the `Option.None`.


To fix it you have 2 options (applying only one will make it works, but I would recommend that you do both to better reflect what you try to achieve):

## Create non empty state

``` csharp
public IState<AuthorizationCredentials> Credentials => State.Value(this, () => new AuthorizationCredentials());
```

## Force the update to set the desired value, no matter current state

```csharp
public async ValueTask SetAuthorizationMode(string mode, CancellationToken ct)
{
	AutorizationMode modeEnum = (AutorizationMode)Enum.Parse(typeof(AutorizationMode), mode);

	await Credentials.Update(creds => (creds ?? new()) with { AutorizationMode = modeEnum }), ct);
}
```

> Note: This is using the `Update()` that will directly unwrap the `Option` for you (providing a `creds = default(AuthorizationCredentials)` if option is not _Some_). You prefer to work with options and `UpdateValue`, you can do : `await Credentials.UpdateValue(opt => (opt.SomeOrDefault() ?? new()) with { AutorizationMode = modeEnum }), ct);`

**Reply by limeniye** [reply]
@dr1rrb, thank you for your answer.
Yes, it works for me.
I am very grateful to you.

**Reply by limeniye** [reply]
@dr1rrb , Yes, it works for me in this case.
But, it works only if all properties inside one type.
For example:
```C#
public partial class MainViewModel
{
    public IState<SomePropery123> A => State.Value(this, () => new SomePropery123());
    public IState<SomePropery123222> B => State.Value(this, () => new SomePropery123222());
    public IState<SomePropery12311111> C => State.Value(this, () => new SomePropery123111());
    public IState<SomePropery44> D => State.Value(this, () => new SomePropery44());
    // Some command
}
```

If I want to split this class into separate entities I have an issue:
```C#
public partial class MainViewModel
{
    public IState<SomeClassA> SomeObjectOne => State.Value(this, () => new SomeClassA());
    public IState<SomeClassB> SomeObjectTwo => State.Value(this, () => new SomeClassB());
}

public record SomeClassA
{
    public IState<SomePropery123> A => State.Value(this, () => new SomePropery123());
    public IState<SomePropery123222> B => State.Value(this, () => new SomePropery123222());    

    // Some command
}

public record SomeClassB
{
    public IState<SomePropery12311111> C => State.Value(this, () => new SomePropery123111());
    public IState<SomePropery44> D => State.Value(this, () => new SomePropery44());
}
```

In view side I changed binding from **="{Binding A.Name}"** to **="{Binding SomeObjectOne.A.Name}"**.

The issue:
```
Error: BindingExpression path error: 'Name' property not found on 'Uno.Extensions.Reactive.Core.StateImpl`1[Ligric.UI.ViewModels.Presentation.ApiDataViewModel]'. BindingExpression: Path='AddingApi.Name' DataItem='Ligric.UI.ViewModels.Presentation.BindableUserApiViewModel'; target element is 'Microsoft.UI.Xaml.Controls.TextBox' (Name='null'); target property is 'Text' (type 'String')
```

The important thing is that this error only appears for IState<T> properties, but not for commands. All commands work correctly.

**Reply by dr1rrb** [reply]
We generate the "bindable friendly proxy" only for classes suffixed by "Model" (configurable in your assembly). So here for the binding path `SomeObjectOne.A` the returned object is an `IState` (not an instance of `SomePropery123`), which does not have any property named `Name`.

But even if you suffix your class by "Model" (and use the generated type `BindableSomeClassAModel`), it will not work properly: entities that are going through feeds are expected to be immutable, which is not the case of `SomeClassA` (as it holds a state which is by nature mutable). If you do something like that, any update on `IState<SomeClassA> SomeObjectOne` will create a new instance of `SomeClassA` driving to possible loose of information.

I would recommend that you keep the `IState` only at the root level of your models (e.g `SomeClassB` / `SomeObjectTwo`) and re-use the entity itself. But if you really want to re-use a sub _model_ between pages (for commands for instance), you should be able to instance the generated proxy directly (be aware that is not yet a officially supported use-case, you might face some weird issues), e.g. `SomeClassAModel` / `SomeObjectOne`:

```csharp
public partial class MainViewModel
{
    public BindableSomeClassAModel SomeObjectOne { get; } = new();
    public IState<SomeClassB> SomeObjectTwo => State.Value(this, () => new SomeClassB());
}

public record SomeClassAModel
{
    public IState<SomePropery123> A => State.Value(this, () => new SomePropery123());
    public IState<SomePropery123222> B => State.Value(this, () => new SomePropery123222());    

    // Some command
}

public record SomeClassB
{
    public SomePropery12311111 C  { get; init; } = new();
    public SomePropery44 D  { get; init; } = new();
}
```
